core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -719,6 +718,7 @@ protected void doRun() throws Exception {
         private void executeLocally() throws Exception {
             // execute locally
             Tuple<Response, ReplicaRequest> primaryResponse = shardOperationOnPrimary(state.metaData(), request);
+            primaryResponse.v2().primaryTerm(indexShardReference.opPrimaryTerm());
```

Comment:
I wonder if the primary term should instead be returned by the call to `shardOperationOnPrimary` above. This would give it the semantics of "I indexed this document when shard had the primary term XYZ". Now there is no clear connection between the two.